src/OmniSharp.Host/MSBuild/Discovery/Providers/StandAloneInstanceProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using Microsoft.Extensions.Logging;
+
+namespace OmniSharp.MSBuild.Discovery.Providers
+{
+    internal class StandAloneInstanceProvider : MSBuildInstanceProvider
+    {
+        public StandAloneInstanceProvider(ILoggerFactory loggerFactory)
+            : base(loggerFactory)
+        {
+        }
+
+        public override ImmutableArray<MSBuildInstance> GetInstances()
+        {
+            var path = FindMSBuildDirectory();
+            if (path == null)
+            {
+                return ImmutableArray<MSBuildInstance>.Empty;
+            }
+
+            var extensionsPath = path;
+            var toolsPath = Path.Combine(extensionsPath, "15.0", "Bin");
+            var roslynPath = Path.Combine(toolsPath, "Roslyn");
+
+            var propertyOverrides = ImmutableDictionary.CreateBuilder<string, string>(StringComparer.OrdinalIgnoreCase);
+            propertyOverrides.Add("MSBuildToolsPath", toolsPath);
+            propertyOverrides.Add("MSBuildExtensionsPath", extensionsPath);
+            propertyOverrides.Add("CscToolPath", roslynPath);
+            propertyOverrides.Add("CscToolExe", "csc.exe");
+
+            return ImmutableArray.Create(
+                new MSBuildInstance(
+                    "STANDALONE",
+                    toolsPath,
+                    new Version(15, 0),
+                    DiscoveryType.StandAlone,
+                    propertyOverrides.ToImmutable(),
+                    setMSBuildExePathVariable: true));
+        }
+
+        private static string FindMSBuildDirectory()
+        {
+            // If OmniSharp is running normally, MSBuild is located in an 'msbuild' folder beneath OmniSharp.exe.
+            var msbuildDirectory = Path.Combine(AppContext.BaseDirectory, "msbuild");
+
+            if (!Directory.Exists(msbuildDirectory))
+            {
+                // If the 'msbuild' folder does not exists beneath "OmniSharp.exe", this is likely a development scenario,
+                // such as debugging or running unit tests. In that case, we try to locate the ".msbuild" folder at the
+                // solution level.
+                msbuildDirectory = FindLocalMSBuildFromSolution();
+            }
+
+            return msbuildDirectory;
+        }
+
+        private static string FindLocalMSBuildFromSolution()
```

Comment:
What about adding a command line argument that we can pass while debugging that specifies the standalone msbuild location? That way we don't have to talk about "OmniSharp.sln" in the product code.

Reply:
Maybe, but where would it get populated from? Note that the product code already referenced "OmniSharp.sln". This PR just moved that code to here.

Reply:
We would add the argument to the arguments passed by the debugger or by tests. Not a huge deal either way.

Reply:
I suppose we could add some MSBuild logic to add an argument passed to the debugger. I'm not sure how it'd work for tests, but I'm guessing we could figure something out. I'm just not sure it's all that valuable.